### PR TITLE
Add Scons

### PR DESCRIPTION
**Pull request by PucklaMotzer09**
Adds scons as a package

### CODE REVIEW

**Comment by PucklaMotzer09** [comment]
Windows doesn't find python3:
```
C:\Users\runneradmin\AppData\Local\.xmake\packages\p\python\3.8.5\bc33c7694ee44a7a872afa7b5dcfbac2\Lib\site-packages\pkg_resources\py2_warn.py:15: UserWarning: Setuptools no longer works on Python 2
************************************************************
Encountered a version of Setuptools that no longer supports
this version of Python. Please head to
https://bit.ly/setuptools-py2-sunset for support.
************************************************************
  warnings.warn(pre + "*" * 60 + msg + "*" * 60)
error: @programdir\core\sandbox\modules\os.lua:388: execv(C:\hostedtoolcache\windows\Python\2.7.18\x86\Python.exe setup.py install --prefix C:\Users\runneradmin\AppData\Local\.xmake\packages\s\scons\4.1.0\d5441f7c8c9a44dc9828b70ce497aa55) failed(32)
```

I think more versions need to be added to ```xmake/modules/detect/tools/find_python3.lua```:
```lua
if is_host("windows") then
        opt.paths = opt.paths or {}
        table.insert(opt.paths, "$(reg HKEY_CURRENT_USER\\SOFTWARE\\Python\\PythonCore\\3.7\\InstallPath;ExecutablePath)")
        table.insert(opt.paths, "$(reg HKEY_CURRENT_USER\\SOFTWARE\\Python\\PythonCore\\3.7-32\\InstallPath;ExecutablePath)")
        table.insert(opt.paths, "$(reg HKEY_CURRENT_USER\\SOFTWARE\\Python\\PythonCore\\3.6\\InstallPath;ExecutablePath)")
        table.insert(opt.paths, "$(reg HKEY_CURRENT_USER\\SOFTWARE\\Python\\PythonCore\\3.6-32\\InstallPath;ExecutablePath)")
end
```

File: packages/s/scons/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    on_load("@linux", "@macosx", "@msys", function (package)
+        import("lib.detect.find_tool")
+
+        -- get version from python
+        local python = assert(find_tool("python3"), "python3 not found!")
+        local py_out = os.iorunv(python.program, {"--version"})
+        local index = py_out:find("%.")
+        local version_major = py_out:sub(index - 1, index - 1)
+        local index1 = py_out:find("%.", index + 1)
+        local version_minor = py_out:sub(index + 1, index1 - 1)
```
Comment (waruqi) [inline packages/s/scons/xmake.lua:23]: we need only use `package:dep("python"):version()

File: packages/s/scons/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        -- set PYTHONPATH
+        local pyver = ("python%d.%d"):format(version_major, version_minor)
+        local PYTHONPATH = package:installdir("lib", pyver, "site-packages", scons_egg)
+        package:addenv("PYTHONPATH", PYTHONPATH)
+    end)
+
+    on_install("@windows", "@linux", "@macosx", "@msys", function (package)
+        import("lib.detect.find_tool")
+
+        local python = assert(find_tool("python3"), "python3 not found!")
```
Comment (waruqi) [inline packages/s/scons/xmake.lua:38]: we need not use find_tool, please run `python` directly

add_deps("python") has bound python env.

File: packages/s/scons/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+        -- setup.py install needs these
+        os.mkdir("build/doc/man")
+        io.writefile("build/doc/man/scons.1", "")
+        io.writefile("build/doc/man/scons-time.1", "")
+        io.writefile("build/doc/man/sconsign.1", "")
+
+        os.execv(python.program, {"setup.py", "install", "--prefix", package:installdir()})
```
Comment (waruqi) [inline packages/s/scons/xmake.lua:46]: os.vrunv("python", ...)

File: packages/s/scons/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        if package:is_plat("windows") then
+            os.mv(package:installdir("Scripts"), package:installdir("bin"))
+        end
+    end)
+
+    on_test(function (package)
+        import("lib.detect.find_tool")
+        local scons = assert(find_tool("scons"), "scons not found!")
```
Comment (waruqi) [inline packages/s/scons/xmake.lua:54]: os.vrun("scons --version")

File: packages/s/scons/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        package:addenv("PYTHONPATH", PYTHONPATH)
+    end)
+
+    on_install("@windows", "@linux", "@macosx", "@msys", function (package)
+        import("lib.detect.find_tool")
+
+        local python = assert(find_tool("python3"), "python3 not found!")
+
+        -- setup.py install needs these
+        os.mkdir("build/doc/man")
```
Comment (waruqi) [inline packages/s/scons/xmake.lua:41]: need not mkdir, io.writefile will mkdir it

File: packages/s/scons/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    set_homepage("https://scons.org")
+    set_description("A software construction tool")
+
+    add_urls("https://github.com/SCons/scons/archive/refs/tags/$(version).zip",
+             "https://github.com/SCons/scons.git")
+    add_versions("4.1.0", "106259e92ba001feae5b50175bcec92306d0420bb08229fb037440cf303fcfc3")
+
+    add_deps("python >=3.0")
+
+    set_kind("binary")
```
Comment (waruqi) [inline packages/s/scons/xmake.lua:12]: Please move to the top

**Comment by PucklaMotzer09** [comment]
I had to move all contents from on_load to on_install, because xmake kept finding the system installed python and package:dep("python") returned nil.

File: packages/s/scons/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package("scons")
+    set_kind("binary")
+
```
Comment (waruqi) [inline packages/s/scons/xmake.lua:3]: remove space line

File: packages/s/scons/xmake.lua [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    set_description("A software construction tool")
+
+    add_urls("https://github.com/SCons/scons/archive/refs/tags/$(version).zip",
+             "https://github.com/SCons/scons.git")
+    add_versions("4.1.0", "106259e92ba001feae5b50175bcec92306d0420bb08229fb037440cf303fcfc3")
+
+    add_deps("python >=3.0")
+
+    on_install("@windows", "@linux", "@macosx", "@msys", function (package)
+        -- get version from python
```
Comment (waruqi) [inline packages/s/scons/xmake.lua:14]: please remove these comments

File: packages/s/scons/xmake.lua (hunk 7 above)
Comment (PucklaMotzer09) [reply packages/s/scons/xmake.lua:3]: What do you mean by "space line"? The indent in line 2? Or should I add a line between package("scons") and set_kind("binary")?

**Comment by waruqi** [comment]
> Windows doesn't find python3:
> 
> ```
> C:\Users\runneradmin\AppData\Local\.xmake\packages\p\python\3.8.5\bc33c7694ee44a7a872afa7b5dcfbac2\Lib\site-packages\pkg_resources\py2_warn.py:15: UserWarning: Setuptools no longer works on Python 2
> ************************************************************
> Encountered a version of Setuptools that no longer supports
> this version of Python. Please head to
> https://bit.ly/setuptools-py2-sunset for support.
> ************************************************************
>   warnings.warn(pre + "*" * 60 + msg + "*" * 60)
> error: @programdir\core\sandbox\modules\os.lua:388: execv(C:\hostedtoolcache\windows\Python\2.7.18\x86\Python.exe setup.py install --prefix C:\Users\runneradmin\AppData\Local\.xmake\packages\s\scons\4.1.0\d5441f7c8c9a44dc9828b70ce497aa55) failed(32)
> ```
> 
> I think more versions need to be added to `xmake/modules/detect/tools/find_python3.lua`:
> 
> ```lua
> if is_host("windows") then
>         opt.paths = opt.paths or {}
>         table.insert(opt.paths, "$(reg HKEY_CURRENT_USER\\SOFTWARE\\Python\\PythonCore\\3.7\\InstallPath;ExecutablePath)")
>         table.insert(opt.paths, "$(reg HKEY_CURRENT_USER\\SOFTWARE\\Python\\PythonCore\\3.7-32\\InstallPath;ExecutablePath)")
>         table.insert(opt.paths, "$(reg HKEY_CURRENT_USER\\SOFTWARE\\Python\\PythonCore\\3.6\\InstallPath;ExecutablePath)")
>         table.insert(opt.paths, "$(reg HKEY_CURRENT_USER\\SOFTWARE\\Python\\PythonCore\\3.6-32\\InstallPath;ExecutablePath)")
> end
> ```

I will improve it later, you can open a issue.

**Pull request merged**